internal/ecies/ecies.go

```diff
@@ -0,0 +1,144 @@
+// This package implements an ECIES (Elliptic Curve Integrated Encryption
+// Scheme) encryption. It uses AES256-GCM for encryption and HKDF with SHA256
+// for key derivation. The package provides functions to encrypt and decrypt
+// messages using a shared secret derived between two parties using ECDH
+// (Elliptic Curve Diffie-Hellman).
+// Inspiration for parts of the code in this package was taken from
+// https://github.com/ecies/go.
+
+package ecies
+
+import (
+	"bytes"
+	"crypto/aes"
+	"crypto/cipher"
+	"crypto/rand"
+	"crypto/sha256"
+	"fmt"
+	"io"
+
+	"golang.org/x/crypto/hkdf"
+)
+
+const (
+	// GCMNonceSize is the size of the nonce used in AES-GCM mode.
+	GCMNonceSize = 16
+)
+
+// EncryptSha256Aes256 encrypts the given message using AES256-GCM with a shared
+// secret (usually derived using ECDH between the sender's ephemeral key and the
+// receiver's public key) that is stretched using HKDF with SHA256. The output
+// is a byte slice containing:
+//
+//	<16 bytes nonce> <16 bytes tag> <... bytes ciphertext>
+func EncryptSha256Aes256(sharedSecret [32]byte, msg []byte) ([]byte, error) {
+	// We begin by stretching the shared secret using HKDF with SHA256.
+	stretchedKey, err := HkdfSha256(sharedSecret[:])
+	if err != nil {
+		return nil, fmt.Errorf("cannot derive hkdf key: %w", err)
+	}
+
+	// Then we create a new AES block cipher using the stretched key. With
+	// the stretched key length of 32 bytes, this will be AES256.
+	block, err := aes.NewCipher(stretchedKey[:])
+	if err != nil {
+		return nil, fmt.Errorf("cannot create new aes block: %w", err)
+	}
+
+	nonce := make([]byte, GCMNonceSize)
+	if _, err := rand.Read(nonce); err != nil {
+		return nil, fmt.Errorf("cannot read random bytes for nonce: "+
+			"%w", err)
+	}
+
+	// The buffer will contain the nonce, tag and ciphertext. We preallocate
+	// the buffer to the final size to avoid multiple allocations.
+	var result bytes.Buffer
+	result.Write(nonce)
+
+	gcm, err := cipher.NewGCMWithNonceSize(block, GCMNonceSize)
```

Comment:
Any particular reason to use this over `NewGCM`? 12-byte nonces are standardized, it's also what [the underlying Go packages uses](https://cs.opensource.google/go/go/+/refs/tags/go1.24.3:src/crypto/cipher/gcm.go;l=19) with the default constructor.

Reply:
See section 5.2.1 of this PDF: https://nvlpubs.nist.gov/nistpubs/legacy/sp/nistspecialpublication800-38d.pdf

> For IVs, it is recommended that implementations restrict support to the length of 96 bits, to
promote interoperability, efficiency, and simplicity of design. 

IV here is basically a nonce.

Reply:
One alternative crypto system to consider is `chacha20poly1305`: https://pkg.go.dev/golang.org/x/crypto/chacha20poly1305#New. This is what we use for encryption in LN. It also natively supports adding associative data (plaintext data included in the MAC).

Reply:
Hmm, I just used an example I found somewhere.
But I like ChaCha20Poly1305 better, changed the code to use that. We'll use the sender's ephemeral public key as the additional data, to authenticate the ciphertext. It will be part of the outer (authmailbox message) envelope, not the ciphertext itself. Does that make sense?